### PR TITLE
[PIR][AutoParallel] Support merge fetch results among micro_scopes

### PR DESCRIPTION
**Pull request by zhaoyinglia**
### PR types

Others
### PR changes

Others
### Description

Pcard-76459
- support  the multi-scope fetch result merge
- add `FetchTensors` to get all multi-scope fetch results
- add `MergeFetchTensors` to merge fetch results of the same name and different scope

### CODE REVIEW

File: paddle/fluid/framework/new_executor/feed_fetch_utils.h [hunk 1]
```diff
@@ -27,10 +27,24 @@ void SetColAttrForFeedFetchOps(std::shared_ptr<ProgramDesc> program_desc,
...
+                      Scope* scope,
+                      std::vector<std::vector<phi::DenseTensor>>* out);
+
+void FetchTensors(const std::vector<std::string>& job_fetch_names,
+                  const std::vector<std::string>& fetch_var_names,
+                  const int64_t micro_batch_id,
+                  Scope* scope,
+                  FetchUnmergedList* fetch_list);
+
+void MergeFetchTensors(const FetchUnmergedList& fetch_list,
```
Comment (From00) [inline paddle/fluid/framework/new_executor/feed_fetch_utils.h:41]:
```suggestion
void MergeFetchList(const FetchUnmergedList& fetch_list,
```

File: paddle/fluid/framework/new_executor/interpreter/job.h [hunk 2]
```diff
@@ -71,11 +54,17 @@ class Job final {
     skip_gc_vars_ = skip_gc_vars;
   }
 
+  void SetFetchVarName(std::string fetch_var_name) {
```
Comment (From00) [inline paddle/fluid/framework/new_executor/interpreter/job.h]:
```suggestion
  void SetFetchVarName(const std::string& fetch_var_name) {
```

File: paddle/fluid/framework/new_executor/pir_interpreter.cc [hunk 3]
```diff
@@ -1100,10 +1101,12 @@ paddle::framework::FetchList PirInterpreter::Run(
...
     framework::FetchList fetch_res;
 
-    for (auto& var_name : fetch_var_names_) {
-      auto* var = inner_scope->FindVar(var_name);
-      VLOG(0) << "fetch " << var_name << "[" << var << "]";
-      fetch_res.push_back(var->Get<phi::DenseTensor>());
+    if (need_fetch) {
+      for (auto& var_name : fetch_var_names_) {
+        auto* var = inner_scope->FindVar(var_name);
+        VLOG(0) << "fetch " << var_name << "[" << var << "]";
```
Comment (From00) [inline paddle/fluid/framework/new_executor/pir_interpreter.cc]:
```suggestion
        VLOG(6) << "fetch " << var_name << "[" << var << "]";
```

File: paddle/fluid/framework/new_executor/feed_fetch_utils.cc [hunk 4]
```diff
@@ -48,42 +49,202 @@ void SetColAttrForFeedFetchOps(std::shared_ptr<ProgramDesc> program_desc,
...
+    int64_t numel_size = feed_tensor.dims()[0];
+    PADDLE_ENFORCE_EQ(numel_size % micro_batch_num,
+                      0,
+                      phi::errors::InvalidArgument(
+                          "Split expects feed data (%s)'s dim[0] (%d) is "
+                          "diviable by micro_batch_num (%d).",
+                          feed_names[i],
+                          numel_size,
+                          micro_batch_num));
+    int64_t split_size = (numel_size + micro_batch_num - 1) / micro_batch_num;
```
Comment (From00) [inline paddle/fluid/framework/new_executor/feed_fetch_utils.cc]:
```suggestion
    int64_t split_size = numel_size / micro_batch_num;
```

File: paddle/fluid/framework/new_executor/feed_fetch_utils.cc [hunk 5]
```diff
@@ -48,42 +49,202 @@ void SetColAttrForFeedFetchOps(std::shared_ptr<ProgramDesc> program_desc,
...
+    VLOG(4) << "Split feed data:" << feed_names[i] << ", dims:("
+            << feed_tensor.dims() << "), micro_batch_num:" << micro_batch_num;
+    for (int64_t j = 0; j < micro_batch_num; ++j) {
+      (*out)[j].resize(i + 1);
```
Comment (From00) [inline paddle/fluid/framework/new_executor/feed_fetch_utils.cc:88]: 可以直接分配feed_tensors.size个元素空间

File: paddle/fluid/framework/new_executor/feed_fetch_utils.h [hunk 6]
```diff
@@ -27,10 +27,24 @@ void SetColAttrForFeedFetchOps(std::shared_ptr<ProgramDesc> program_desc,
...
+                       const int64_t micro_batch_num,
+                       FetchList* out);
+
+void MergeTensors(const std::vector<const phi::DenseTensor*>& tensors,
+                  platform::Place dst_place,
```
Comment (From00) [inline paddle/fluid/framework/new_executor/feed_fetch_utils.h]:
```suggestion
                  const platform::Place& dst_place,
```

**Review by From00 — DISMISSED** [review]
LGTM overall

File: paddle/fluid/framework/new_executor/feed_fetch_utils.h (hunk 6 above)
Comment (zhaoyinglia) [reply paddle/fluid/framework/new_executor/feed_fetch_utils.h]: Done.

File: paddle/fluid/framework/new_executor/feed_fetch_utils.cc (hunk 4 above)
Comment (zhaoyinglia) [reply paddle/fluid/framework/new_executor/feed_fetch_utils.cc]: Done.

File: paddle/fluid/framework/new_executor/pir_interpreter.cc (hunk 3 above)
Comment (zhaoyinglia) [reply paddle/fluid/framework/new_executor/pir_interpreter.cc]: Done.

File: paddle/fluid/framework/new_executor/interpreter/job.h (hunk 2 above)
Comment (zhaoyinglia) [reply paddle/fluid/framework/new_executor/interpreter/job.h]: Done.

**Review by From00 — APPROVED** [review]

**Review by zhiqiu — APPROVED** [review]
LGTM

**Pull request merged**